### PR TITLE
LSP Rewrite Step 1: fix many bugs, more LSP test coverage!

### PR DESCRIPTION
**Pull request by acao**
this fixes multiple cacheing bugs, on writing some in-depth integration coverage for the LSP server.
it also solves several bugs regarding loading config types, and properly restarts the server when there are config changes

### Bugfix Summary

- jump to definition in embedded files offset bug (correct file, wrong range)
- cache invalidation for fragments (fragment definitions change when they are supposed to)
- schema cache invalidation for schema files (schema doesn't change when a watched schema file changes)
- schema definition lookups & autocomplete crossing into the wrong workspace (project name wasn't being resolved when building the cache keys! literally just one line!)
- type validation works in more places it didn't work before because of the fixed fragment and schema type cache bugs!
- autocompletion for SDL type fields and input type fields is restored! 🎉 
- the lsp server only reloads the config when/if package.json contains a `graphql` entry

### Known Bugs Fixed
- #3318
- #2357
- #3469
- #2422 
- #2820
- #3538
- many others to add here...

### Test Improvements

- new, high level integration spec suite for the LSP with a matching test utility
- more unit test coverage
- **total increased test coverage of about 23% in the LSP server codebase.**
- many "happy paths" covered for both schema and code first contexts
- many bugs revealed (and their source)

### What Happens After This?:

- full schema lifecycle for "code first"/ cached schema file context (i.e. schema from URL updates when it changes and re-saves the file, at least with a configurable timeout)

### Is there a Prerelease? <a href="#user-content-prerelease" id="prerelease">#</a>

Try it out here! Just install the .vsix in the .zip. If you still have any issues, please report them in the applicable bug ticket, noting there is still an issue with the new 0.10.x release. We will try to capture all edge cases in the new test suite

https://github.com/graphql/graphiql/files/14475534/vscode-bugfixes.zip

### Test methodology thoughts

overall, this new quasi behavioral integration spec plane serves as a learning tool for contributors, a refactoring tool (I may have a stash for a new chained branch ready to go haha) and a tool for discovering bugs and readily identifying their sources, even allowing us to use a _BDD_ approach to fixing bugs (and soon adding features!) 😍  !!

the only major caveat of this testing approach is that, like with many high level, quasi-e2e integration tests, these rely on our assumptions of how LSP clients are using our service interfaces. This is something that changes ever so slightly and even in semi-breaking ways in vscode (like with the introduction of multi-root workspaces), so careful readings of the `vscode-languageserver` changelog and probably also the LSP protocol spec will be required to maintain this spec, and not cause mysterious false positives

### CODE REVIEW

**Comment by acao** [comment]
For anyone who wants to give it a try, here is a zip of the 0.10.0 vsix!
[vscode-bugfixes.zip](https://github.com/graphql/graphiql/files/14475534/vscode-bugfixes.zip)

File: .changeset/rotten-seahorses-fry.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+---
+'graphql-language-service-server': minor
+'vscode-graphql': minor
+'graphql-language-service-server-cli': minor
+---
+
+Fix many schema and fragment lifecycle issues, for all contexts except for schema updates for url schemas.
+Note: this makes `cacheSchemaForLookup` enabled by default again for schema first contexts.
+
+this fixes multiple cacheing bugs, on writing some in-depth integration coverage for the LSP server.
```
Comment (TallTed) [inline .changeset/rotten-seahorses-fry.md:10]:
```suggestion
this fixes multiple caching bugs, on writing some in-depth integration coverage for the LSP server.
```

File: .changeset/rotten-seahorses-fry.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+---
+'graphql-language-service-server': minor
+'vscode-graphql': minor
+'graphql-language-service-server-cli': minor
+---
+
+Fix many schema and fragment lifecycle issues, for all contexts except for schema updates for url schemas.
```
Comment (TallTed) [inline .changeset/rotten-seahorses-fry.md:7]:
```suggestion
Fix many schema and fragment lifecycle issues, for all contexts except schema updates of url schemas.
```

File: .changeset/rotten-seahorses-fry.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+- #3469
+- #2422
+- #2820
+- many others to add here...
+
+### Test Improvements
+
+- new, high level integration spec suite for the LSP with a matching test utility
+- more unit test coverage
+- **total increased test coverage of about 25% in the LSP server codebase.**
```
Comment (TallTed) [inline .changeset/rotten-seahorses-fry.md:33]:
```suggestion
- **total increased test coverage of about 25% in the LSP server codebase**
```

**Review by TallTed — CHANGES_REQUESTED** [review]
small, human-facing

File: packages/graphql-language-service-server/src/MessageProcessor.ts [hunk 4]
```diff
@@ -1041,26 +1076,33 @@ export class MessageProcessor {
...
+      'generated-schema.graphql',
     );
+    const fsPath = this._getTmpProjectPath(
+      project,
+      false,
+      'generated-schema.graphql',
+    );
+    // invalidate the cache for the generated schema file
+    // whether or not the user will continue using this feature
+    // because sdlOnly needs this file to be removed as well if the user is switching schemas
```
Comment (TallTed) [inline packages/graphql-language-service-server/src/MessageProcessor.ts:1097]:
```suggestion
    // because sdlOnly also needs this file to be removed, if the user is switching schemas
```

File: packages/vscode-graphql/README.md [hunk 5]
```diff
@@ -66,7 +68,13 @@ export default {
 For more help with configuring the language server,
 [the language server readme](https://github.com/graphql/graphiql/tree/main/packages/graphql-language-service-server/README.md)
 is the source of truth for all settings used by all editors which use the
-language server. The [`graphql-config`](https://graphql-config.com/) docs are also very helpful.
+language server.
+
+This includes LSP settings provided by extensions like `vscode-graphql`, nvim, etc.
+
+There are a number of configurations that can be provided from both editor settings or the graphql config file, and the editor setting takes precedence, to allow users to override their graphql config file settings in a user context as needed.
```
Comment (TallTed) [inline packages/vscode-graphql/README.md:75]:
```suggestion
There are a number of configurations that can be provided from editor settings and/or the graphql config file, with the editor settings taking precedence, so users can override their graphql config file settings in a user context, as needed.
```

File: packages/vscode-graphql/package.json [hunk 6]
```diff
@@ -107,6 +107,13 @@
           "description": "Fail the request on invalid certificate",
           "default": true
         },
+        "vscode-graphql.schemaCacheTTL": {
+          "type": [
+            "number"
+          ],
+          "description": "Schema cache ttl in milliseconds before requesting a fresh schema when caching the local schema file is enabled. Defaults to 30000 (30 seconds).",
```
Comment (TallTed) [inline packages/vscode-graphql/package.json]:
```suggestion
          "description": "Schema cache ttl, in milliseconds before requesting a fresh schema when caching the local schema file is enabled. Defaults to 30000 (30 seconds).",
```
